README.md
[![banner](https://raw.githubusercontent.com/oceanprotocol/art/master/github/repo-banner%402x.png)](https://oceanprotocol.com)

# squid-py

> 💧 Python wrapper, allowing to integrate the basic Ocean/web3.py capabilities
> [oceanprotocol.com](https://oceanprotocol.com)

[![PyPI](https://img.shields.io/pypi/v/squid-py.svg)](https://pypi.org/project/squid-py/)
[![Travis (.com)](https://img.shields.io/travis/com/oceanprotocol/squid-py.svg)](https://travis-ci.com/oceanprotocol/squid-py)
[![GitHub contributors](https://img.shields.io/github/contributors/oceanprotocol/squid-py.svg)](https://github.com/oceanprotocol/squid-py/graphs/contributors)
[![Codacy Badge](https://img.shields.io/codacy/grade/7084fbf528934327904a49d458bc46d1.svg)](https://app.codacy.com/project/ocean-protocol/squid-py/dashboard)
[![Codacy coverage](https://img.shields.io/codacy/coverage/7084fbf528934327904a49d458bc46d1.svg)](https://app.codacy.com/project/ocean-protocol/squid-py/dashboard)

---

## Table of Contents

  - [Features](#features)
  - [Prerequisites](#prerequisites)
  - [Quick-start](#quick-start)
     - [Usage:](#usage)
  - [Configuration](#configuration)
  - [Development](#development)
        - [Code style](#code-style)
        - [Testing](#testing)
  - [License](#license)

---

## Features

Squid-py include the methods to make easy the connection with contracts deployed in different networks.
This repository include also the methods to encrypt and decrypt information using the Parity Secret Store.

## Prerequisites

Python 3.6

## Quick-start

Install Squid:

```
pip install squid-py
```

### Usage:

```python
import os
import time

from squid_py import (
    Ocean,
    ConfigProvider,
    Config,
    Metadata,
    Account
)

ConfigProvider.set_config(Config('config.ini'))
# Make a new instance of Ocean
ocean = Ocean()
config = ConfigProvider.get_config()
# make account instance, assuming the ethereum account and password are set 
# in the config file `config.ini`
account = ocean.accounts.list()[0]
# or 
account = Account(config.parity_address, config.parity_password)

# PUBLISHER
# Let's start by registering an asset in the Ocean network
metadata = Metadata.get_example()

# purchase and service endpoints require `brizo.url` is set in the config file
# or passed to Ocean instance in the config_dict.
# define the services to include in the new asset DDO

ddo = ocean.assets.create(metadata, account)

# Now we have an asset registered, we can verify it exists by resolving the did
_ddo = ocean.assets.resolve(ddo.did)
# ddo and _ddo should be identical

# CONSUMER
# search for assets
asset_ddo = ocean.assets.search('Ocean protocol')[0]
# Need some ocean tokens to be able to purchase assets
ocean.accounts.request_tokens(account, 10)

# Start the purchase/consume request. This will automatically make a payment from the specified account.
consumer_account = ocean.accounts.list()[1]
service_agreement_id = ocean.assets.order(asset_ddo.did, 0, consumer_account)

# after a short wait (seconds to minutes) the asset data files should be available in the `downloads.path` defined in config
# wait a bit to let things happen
time.sleep(20)

# Asset files are saved in a folder named after the asset id
dataset_dir = os.path.join(ConfigProvider.get_config().downloads_path, f'datafile.{asset_ddo.asset_id}.0')
if os.path.exists(dataset_dir):
    print('asset files downloaded: {}'.format(os.listdir(dataset_dir)))

```

## Configuration

```python
config_dict = {
    'keeper-contracts': {
        # Point to an Ethereum RPC client. Note that Squid learns the name of the network to work with from this client.
        'keeper.url': 'http://localhost:8545',
        # Specify the keeper contracts artifacts folder (has the smart contracts definitions json files). When you
        # install the package, the artifacts are automatically picked up from the `keeper-contracts` Python
        # dependency unless you are using a local ethereum network.
        'keeper.path': 'artifacts',
        'secret_store.url': 'http://localhost:12001',
        'parity.url': 'http://localhost:8545',
        'parity.address': '',
        'parity.password': '',

    },
    'resources': {
        # aquarius is the metadata store. It stores the assets DDO/DID-document
        'aquarius.url': 'http://localhost:5000',
        # Brizo is the publisher's agent. It serves purchase and requests for both data access and compute services
        'brizo.url': 'http://localhost:8030',
        # points to the local database file used for storing temporary information (for instance, pending service agreements).
        'storage.path': 'squid_py.db',
        # Where to store downloaded asset files
        'downloads.path': 'consume-downloads'
    }
}

```

In addition to the configuration file, you may use the following environment variables (override the corresponding configuration file values):

- KEEPER_PATH
- KEEPER_URL
- GAS_LIMIT
- AQUARIUS_URL

## Development

1. Set up a virtual environment

    ```bash
    virtualenv venv -p python3.6
    source venv/bin/activate 
    ```

1. Install requirements

    ```
    pip install -r requirements_dev.txt
    ```

1. Create the local testing environment using [barge](https://github.com/oceanprotocol/barge). Once cloned that repository, you can start the cluster running:

    ```
    ./start_ocean.sh --latest --no-brizo --no-pleuston --local-spree-node
    ```

    It runs an Aquarius node and an Ethereum RPC client. For details, read `docker-compose.yml`.

1. Create local configuration file

    ```
    cp config.ini config_local.ini
    ```

   `config_local.ini` is used by unit tests.

1. Copy keeper artifacts

    A bash script is available to copy keeper artifacts into this file directly from a running docker image. This script needs to run in the root of the project.
    The script waits until the keeper contracts are deployed, and then copies the artifacts.

    ```
    ./scripts/wait_for_migration_and_extract_keeper_artifacts.sh
    ```

    The artifacts contain the addresses of all the deployed contracts and their ABI definitions required to interact with them.

1. Run the automated tests

    ```
    TEST_TIER=<tier> python3 setup.py test # where <tier> is unit, integration or e2e, depending on how much of the tests you want to run (e2e runs everything)
    ```

1. Run tests automatically as you change code while doing TDD

    ```
    TEST_TIER=<tier> ptw
    ```

#### Code style

The information about code style in python is documented in this two links [python-developer-guide](https://github.com/oceanprotocol/dev-ocean/blob/master/doc/development/python-developer-guide.md)
and [python-style-guide](https://github.com/oceanprotocol/dev-ocean/blob/master/doc/development/python-style-guide.md).

#### Testing

Automatic tests are setup via Travis, executing `tox`.
Our test use pytest framework.

#### New Version / New Release

See [RELEASE_PROCESS.md](RELEASE_PROCESS.md)

## License

```
Copyright 2018 Ocean Protocol Foundation Ltd.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
History
=======

0.6.13 (Jun 22nd, 2019)
-----------------------
* Pick up always last version of contracts.
* Don't allow dispenser in pacific

0.6.12 (Jun 21st, 2019)
-----------------------
* Fix processing events related to service agreements
* Fix release reward after access is granted
* Increase test coverage

0.6.11 (Jun 14th, 2019)
----------------------
* Fix issue in condition method.
* Fix problem with brizo mock.

0.6.10 (Jun 11th, 2019)
----------------------
* Allow separate config per ocean instance
* Fix issue in timeout processing in multi event listener
* Refactor registering of agreement events watchers
* Add interface and implementation to support provider handling of new agreement events (in squid, caller does not need to implement own handler)

0.6.9 (Jun 6th, 2019)
----------------------
* Remove waiting for transaction receipt when handling agreement events.

0.6.8 (Jun 5th, 2019)
----------------------
* Eliminate processing the same event repeatedly
* Fix a database issue

0.6.7 (May 29th, 2019)
----------------------
* Fix issue in the signature.

0.6.6 (May 28th, 2019)
----------------------
* Begin checksum with 0x.
* Fix small bugs.
* Add version info of the contracts.
* Manage price as a string.

0.6.5 (May 16th, 2019)
----------------------
* Add ocean auth module to support handling signed token.

0.6.4 (May 7th, 2019)
----------------------
* Use the proper version of keeper-contracts that supports consumer creating service agreement.

0.6.3 (April 30th, 2019)
------------------------
* Update `assets.order` to use `agreements.create` instead of `agreements.send`

0.6.2 (April 30th, 2019)
------------------------
* Revert change that was breaking compatibility

0.6.1 (April 30th, 2019)
------------------------
* The agreement can be created by the publisher and the consumer
* Allow to consume individual assets using the index
* Fix some mistakes in the ddo generation
* Improve logging

0.6.0 (April 24th, 2019)
------------------------
* Get list of the Assets Published by an user
* Get list of the Assets Purchased by an user
* Get the SEA status
* Get the Owner of an Asset
* Expose an additional consume interface allowing to use index parameter
* Consumer can initialize a SEA on-chain
  
0.5.14 (April 5th, 2019)
-------------------------
* Update chunk_size in the requests.
* Page should be equal or bigger than 1.

0.5.13 (April 5th, 2019)
-------------------------
* Working with the new Response for queries in aquarius

0.5.12 (April 5th, 2019)
-------------------------
* Custom httpProvider to manage better the connections pool.

0.5.11 (April 1st, 2019)
-------------------------
* Fix to solve an issue handling the signature in web3.
* Prepare changes to make Brizo work with squid-js

0.5.10 (March 29th, 2019)
-------------------------
* Upgrade to keeper contracts 0.9.0
* Update tests and examples to the requirements in the files section of the ddo.

0.5.9 (March 27th, 2019)
-------------------------
* New event handler.
* Keep the files in the ddo.

0.5.8 (March 26th, 2019)
-------------------------
* Upgrade to keeper-contracts 0.8.8.

0.5.7 (March 25th, 2019)
-------------------------
* Unify authentication and publicKey parameters in the ddo.

0.5.6 (March 22nd, 2019)
-------------------------
* Upgrade to the changes in the did registry contract in version 0.8.7.

0.5.5 (March 21st, 2019)
-------------------------
* Fix issue with the validation of the signature.
* Add docstrings missed.

0.5.4 (March 20th, 2019)
-------------------------
* Use personal_sendTransaction to execute smart contract functions to avoid having to unlock account before each transaction.
* Improved error handling.
* Better control of http requests connection pool.

0.5.3 (March 18th, 2019)
------------------------
* Update to keeper 0.8.6

0.5.2 (March 14th, 2019)
-------------------------
* This release works with keeper v0.8.5 which automatically propses/approves the agreement template. Brizo and squid-py should work without the need for worring about approving the agreement template.

0.5.1 (March 13th, 2019)
-------------------------
* Small bug fixed.

0.5.0 (March 6th, 2019)
-------------------------
* Working with keeper-contracts v0.8.1.
* Update with the new squid API.

0.4.4 (February 21st, 2019)
-------------------------
* Change log level in the event listener.

0.4.3 (February 19th, 2019)
-------------------------
* Unlock missed account to avoid Nile error

0.4.2 (February 7th, 2019)
-------------------------
* Secret store bug fixed

0.4.1 (February 6th, 2019)
-------------------------
* Include template in the distribution

0.4.0 (February 6th, 2019)
-------------------------
* This release implements the latest v0.1 squid-api.

0.3.3 (February 6th, 2019)
-------------------------
* Create checksum in the register_asset method

0.3.2 (January 30th, 2019)
-------------------------
* Fixing in the secret-store
* Remove circular dependencies

0.3.1 (January 30th, 2019)
-------------------------
* Support changes in OEP-8

0.3.0 (January 29th, 2019)
-------------------------
* Enable squid to run with keeper-contracts v0.6.x

0.2.24 (January 21st, 2019)
---------------------------
* Remove dids in the services

0.2.21 (January 15th, 2019)
---------------------------
* Fixes a bug that breaks resolving assets in brizo.

0.2.17 (December 18th, 2019)
---------------------------
* Unify DID format across different languages.
* Move unlock_account to the Account class
* Detect existing service agreement and raise error before executing the agreement on-chain
* Raise error in execute_service_agreement when consumer signature is invalid to allow Brizo to provide more accurate error message
* Simplify imports

0.2.14 (December 7th, 2019)
---------------------------
* New docker images need this update.

0.2.11 (November 28th, 2019)
---------------------------
Progress in Service Level Agreements and consume workflow.

This version is still to be integration tested.


squid_py/config.py
"""Config data."""

#  Copyright 2018 Ocean Protocol Foundation
#  SPDX-License-Identifier: Apache-2.0

import configparser
import logging
import os
import site
from pathlib import Path

DEFAULT_KEEPER_HOST = 'localhost'
DEFAULT_KEEPER_PORT = 8545
DEFAULT_KEEPER_URL = 'http://localhost:8545'
DEFAULT_KEEPER_PATH = 'artifacts'
DEFAULT_GAS_LIMIT = 4000000
DEFAULT_NAME_AQUARIUS_URL = 'http://localhost:5000'
DEFAULT_STORAGE_PATH = 'squid_py.db'

NAME_KEEPER_URL = 'keeper.url'
NAME_KEEPER_PATH = 'keeper.path'
NAME_GAS_LIMIT = 'gas_limit'
NAME_AQUARIUS_URL = 'aquarius.url'
NAME_STORAGE_PATH = 'storage.path'

NAME_SECRET_STORE_URL = 'secret_store.url'
NAME_PARITY_URL = 'parity.url'
NAME_PARITY_ADDRESS = 'parity.address'
NAME_PARITY_PASSWORD = 'parity.password'

environ_names = {
    NAME_KEEPER_URL: ['KEEPER_URL', 'Keeper URL'],
    NAME_KEEPER_PATH: ['KEEPER_PATH', 'Path to the keeper contracts'],
    NAME_GAS_LIMIT: ['GAS_LIMIT', 'Gas limit'],
    NAME_AQUARIUS_URL: ['AQUARIUS_URL', 'Aquarius URL'],
    NAME_STORAGE_PATH: ['STORAGE_PATH', 'Path to the local database file'],
    NAME_SECRET_STORE_URL: ['SECRET_STORE_URL', 'Secret Store URL'],
    NAME_PARITY_URL: ['PARITY_URL', 'Parity URL'],
    NAME_PARITY_ADDRESS: ['PARITY_ADDRESS', 'Parity address'],
    NAME_PARITY_PASSWORD: ['PARITY_PASSWORD', 'Parity password'],
}

config_defaults = {
    'keeper-contracts': {
        NAME_KEEPER_URL: DEFAULT_KEEPER_URL,
        NAME_KEEPER_PATH: DEFAULT_KEEPER_PATH,
        NAME_GAS_LIMIT: DEFAULT_GAS_LIMIT,
        NAME_SECRET_STORE_URL: '',
        NAME_PARITY_URL: '',
        NAME_PARITY_ADDRESS: '',
        NAME_PARITY_PASSWORD: '',
    },
    'resources': {
        NAME_AQUARIUS_URL: DEFAULT_NAME_AQUARIUS_URL,
        NAME_STORAGE_PATH: DEFAULT_STORAGE_PATH
    }
}


class Config(configparser.ConfigParser):
    """Class to manage the squid-py configuration."""

    def __init__(self, filename=None, options_dict=None, **kwargs):
        """
        Initialize Config class.

        Options available:

        [keeper-contracts]
        keeper.url = http://localhost:8545                            # Keeper-contracts url.
        keeper.path = artifacts                                       # Path of json abis.
        secret_store.url = http://localhost:12001                     # Secret store url.
        parity.url = http://localhost:8545                            # Parity client url.
        parity.address = 0x00bd138abd70e2f00903268f3db08f2d25677c9e   # Partity account address.
        parity.password = node0                                       # Parity account password.
        [resources]
        aquarius.url = http://localhost:5000                          # Aquarius url.
        brizo.url = http://localhost:8030                             # Brizo url.
        storage.path = squid_py.db                                    # Path of sla back-up storage.

        :param filename: Path of the config file, str.
        :param options_dict: Python dict with the config, dict.
        :param kwargs: Additional args. If you pass text, you have to pass the plain text
        configuration.
        """
        configparser.ConfigParser.__init__(self)

        self.read_dict(config_defaults)
        self._web3_provider = None
        self._section_name = 'keeper-contracts'
        self._logger = logging.getLogger('config')

        if filename:
            self._logger.debug(f'Config: loading config file {filename}')
            with open(filename) as fp:
                text = fp.read()
                self.read_string(text)
        else:
            if 'text' in kwargs:
                self.read_string(kwargs['text'])

        if options_dict:
            self._logger.debug(f'Config: loading from dict {options_dict}')
            self.read_dict(options_dict)

        self._load_environ()

    def _load_environ(self):
        for option_name, environ_item in environ_names.items():
            value = os.environ.get(environ_item[0])
            if value is not None:
                self._logger.debug(f'Config: setting environ {option_name} = {value}')
                self.set(self._section_name, option_name, value)

    @property
    def keeper_path(self):
        """Path where the keeper-contracts artifacts are allocated."""
        keeper_path_string = self.get(self._section_name, NAME_KEEPER_PATH)
        path = Path(keeper_path_string).expanduser().resolve()
        # TODO: Handle the default case and make default empty string
        # assert path.exists(), "Can't find the keeper path: {} ({})"..format(keeper_path_string,
        # path)
        if os.path.exists(path):
            pass
        elif os.getenv('VIRTUAL_ENV'):
            path = os.path.join(os.getenv('VIRTUAL_ENV'), 'artifacts')
        else:
            path = os.path.join(site.PREFIXES[0], 'artifacts')
        return path

    @property
    def storage_path(self):
        """Path to save the current execution of the service agreements and restart if needed."""
        return self.get('resources', NAME_STORAGE_PATH)

    @property
    def keeper_url(self):
        """URL of the keeper. (e.g.): http://mykeeper:8545."""
        return self.get(self._section_name, NAME_KEEPER_URL)

    @property
    def gas_limit(self):
        """Ethereum gas limit."""
        return int(self.get(self._section_name, NAME_GAS_LIMIT))

    @property
    def aquarius_url(self):
        """URL of aquarius component. (e.g.): http://myaquarius:5000."""
        return self.get('resources', NAME_AQUARIUS_URL)

    @property
    def secret_store_url(self):
        """URL of the secret store component. (e.g.): http://mysecretstore:12001."""
        return self.get(self._section_name, NAME_SECRET_STORE_URL)

    @property
    def parity_url(self):
        """URL of parity client. (e.g.): http://myparity:8545."""
        return self.get(self._section_name, NAME_PARITY_URL)

    @property
    def parity_address(self):
        """Parity address. (e.g.): 0x00bd138abd70e2f00903268f3db08f2d25677c9e."""
        return self.get(self._section_name, NAME_PARITY_ADDRESS)

    @property
    def parity_password(self):
        """Parity password for your address. (e.g.): Mypass."""
        return self.get(self._section_name, NAME_PARITY_PASSWORD)

    @property
    def downloads_path(self):
        """Path for the downloads of assets."""
        return self.get('resources', 'downloads.path')

    @property
    def auth_token_message(self):
        return self.get('resources', 'auth_token_message')

    @property
    def auth_token_expiration(self):
        return self.get('resources', 'auth_token_expiration')

    @property
    def web3_provider(self):
        """Web3 provider"""
        return self._web3_provider

    @web3_provider.setter
    def web3_provider(self, web3_provider):
        self._web3_provider = web3_provider
